src/backend/distributed/utils/maintenanced.c

```diff
@@ -257,6 +263,13 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		CHECK_FOR_INTERRUPTS();
 
 		/*
+		 * We clear the metadata cache before every iteration because otherwise
+		 * it might contain stale OIDs if deadlock detection runs during a DROP
+		 * EXTENSION followed by a CREATE EXTENSION.
+		 */
+		ClearMetadataCache();
```

Comment:
Sure, that's what `MetadataTableAccessible` solves for `pg_dist_node` and `pg_dist_local_group`. We make sure we can lock and open the relation prior to actually using it, and if not we skip distributed deadlock detection.

However, `pg_dist_partition` is never opened in the background worker, we only call `get_relname_relid` to get the OID and then expect to later get an invalidation for that OID when the extension is dropped. That assumption doesn't hold, which means our cache gets stale.